### PR TITLE
Feature/binance chain integration

### LINKED ISSUE #7730: Binance chain integration (bounties/hackathons)
### Why Is this Needed

*Summary*:
This integration is to allow users on the platform use the BNB token for 
* Bounties 
* Hackathons

### Description

*Type*: Feature

### Current Behavior
No existing implementation. Existing chains are BTC, ETH, ETC, Celo, Filecoin, Zilliqa, and Polkadot

### Expected Behavior
**Bounties**
* Funder will be able to fund all types of bounties (traditional, cooperative, & contest) using BNB token
* Funder will be able payout to the contributor
* Funder will be prompted by the browser extension to sign the transaction if the funder has enough funds in the wallet
* Once completed -> Gitcoin will check the chain to find the transaction using the block explorer API and place it in a pending state until it the transaction has succeeded (this would be done via polling the chain for transaction status)
* Upon completion of payment funder can close the bounty 
* If the bounty is contest / cooperative -> funders will be able to pay out multiple users in a similar process as mentioned above
* Contributors will be able to enter BNB address to which the tokens would be sent to once the funder approves and does the payout 

**Hackathons**
* Hackathon will be built on top of the bounty integration allowing Binance to host hackathons which can be funded using BNB token allowing multiple users to work on the bounty at the same point in time (again this depends on the what the funder wants)
* Circle integration / Hackathon Town Square for dedicated Binance hackathons  
* Dashboard to view all submitted projects. 

### Definition of Done
[comment]: # (Any other information that would be useful, bullets are helpful.)

**Backend**
- [x] Add `binance_ext` as a payout type in BountyFulfillment model
- [x] Write backend function to fetch transaction status from a binance explorer (mainnet & testnet)
- [x] Write backend function to sync binance payouts
- [x] Add management command to sync binance payout
- [x] Allow `binance_ext` in bounty fulfil and payout views

**Frontend**
- [x] Connect to binance network via binance wallet browser extension
- [x] Fetch connected wallet address and balance
- [x] Sign and submit transactions to the binance network
- [x] Select binance chain while creating a new bounty
- [x] Wait for BinanceChain to be injected into window object before waiting for events

### Data Requirements
One or more successful BNB bounties, yea?

### Additional Information
* Binance chain explorer to sync payouts:
**Mainnet:** https://bscscan.com/
**Testnet:**  https://testnet.bscscan.com/

* Binance Wallet:
**API:** https://docs.binance.org/smart-chain/wallet/wallet_api.html
**Extension:** https://docs.binance.org/smart-chain/wallet/binance.html

### PR DESCRIPTION
**Pull request by chibie**
##### Description

This integration is to allow users on the platform use the BNB token for Bounties and Hackathons

**Backend**
- [x] Add `binance_ext` as a payout type in BountyFulfillment model
- [x] Write backend function to fetch transaction status from a binance explorer (mainnet & testnet)
- [x] Write backend function to sync binance payouts
- [x] Add management command to sync binance payout
- [x] Allow `binance_ext` in bounty fulfil and payout views

**Frontend**
- [x] Connect to binance network via binance wallet browser extension
- [x] Fetch connected wallet address and balance
- [x] Sign and submit transactions to the binance network
- [x] Select binance chain while creating a new bounty
- [x] Wait for BinanceChain to be injected into window object before waiting for events

##### Refers/Fixes

#7730 

##### Testing

- [x] Create Bounty with Binance Chain
- [x] Send payment and confirm
- [x] Sync works

https://drive.google.com/file/d/1dcZpFeLXdg3d58p3f0m-OnXEQoUyipK2/view?usp=sharing

### CODE REVIEW

File: app/dashboard/sync/binance.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        logger.error(f'error: get_binance_txn_status - {e}')
+
+    finally:
+        return response
+
+
+def sync_binance_payout(fulfillment):
+    if fulfillment.payout_tx_id:
+        txn_status = get_binance_txn_status(fulfillment)
+        if txn_status:
```
Comment (danlipert) [inline app/dashboard/sync/binance.py:60]: Here you can save a few lookups I think:

```
txn_status = get_binance_txn_status(fulfillment) 
status_description = txn_status.get('status')
if status_description == 'done':
...
elif status_description == 'expired':
...
```

Just avoids a few .get lookups

**Review by danlipert — CHANGES_REQUESTED** [review]
Looking great so far!

File: app/dashboard/sync/binance.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
+import logging
+
+from django.utils import timezone
+
+import requests
+from dashboard.sync.helpers import record_payout_activity
+
+logger = logging.getLogger(__name__)
+
+headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0. 2272.118 Safari/537.36.'}
```
Comment (thelostone-mc) [inline app/dashboard/sync/binance.py:10]: Curious -> do they block us if we don't have a useragent ?

File: app/dashboard/sync/binance.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        result = binance_response['result']
+
+        response = { 'status': 'pending' }
+
+        if result:
+            tx_status = int(result.get('status'), 16) # convert hex to decimal
+
+            if tx_status == '1':
+                response = { 'status': 'done' }
+            elif tx_status == '0':
```
Comment (thelostone-mc) [inline app/dashboard/sync/binance.py:46]: Is the status when a txn has failed ?
Is there a way to know if txn is pending ?

File: app/dashboard/sync/binance.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+            if status_description == 'done':
+                fulfillment.payout_status = 'done'
+                fulfillment.accepted_on = timezone.now()
+                fulfillment.accepted = True
+                record_payout_activity(fulfillment)
+            elif status_description == 'expired':
+                fulfillment.payout_status = 'expired'
+
+        fulfillment.save()
```
Comment (thelostone-mc) [inline app/dashboard/sync/binance.py:71]: let's push this within the if and else statement !  
cause if status_description is neither done / expired -> we'll keeping saving

**Review by thelostone-mc — CHANGES_REQUESTED** [review]
Could we also update the views.py so that this will be complete entirely + did we test the backend sync_payout for binance ?

https://github.com/gitcoinco/web/blob/master/app/dashboard/views.py#L5589
https://github.com/gitcoinco/web/blob/master/app/dashboard/views.py#L5704
https://github.com/gitcoinco/web/blob/master/app/dashboard/views.py#L5770

File: app/dashboard/sync/binance.py (hunk 2 above)
Comment (chibie) [reply app/dashboard/sync/binance.py:10]: i don't think so. i didn't find any info on that. do we stick to `Host: gitcoin.co` ?

here's their gist on API usage

```
The Binance Smart Chain Developer APIs are provided as a community service and without warranty, so please use what you need and no more. We support both GET/POST requests and there is a rate limit of 5 calls per sec/IP.

Note: Source attribution via a link back or mention that your app is "Powered by BscScan.com APIs" is required except for personal/private usage.
```

File: app/dashboard/sync/binance.py (hunk 3 above)
Comment (chibie) [reply app/dashboard/sync/binance.py:46]: status only captures failure (0x1) or success (0x0)

it's basically eth under the hood.

**Comment by chibie** [comment]
@danlipert @thelostone-mc

since bsc is basically a fork of eth, should we be concerned about replaced transactions?

ps:  i'll write unit tests for binance sync

File: app/dashboard/sync/binance.py (hunk 2 above)
Comment (thelostone-mc) [reply app/dashboard/sync/binance.py:10]: Yup let's stick to that cause that way if we run into trouble and get it resolved!

File: app/dashboard/views.py [hunk 5]
```diff
@@ -5586,7 +5586,7 @@ def fulfill_bounty_v1(request):
     if payout_type == 'fiat' and not fulfiller_identifier:
         response['message'] = 'error: missing fulfiller_identifier'
         return JsonResponse(response)
-    elif (payout_type == 'qr' or payout_type == 'polkadot-ext') and not fulfiller_address:
+    elif payout_type in ['qr', 'polkadot-ext', 'binance-ext'] and not fulfiller_address:
```
Comment (thelostone-mc) [inline app/dashboard/views.py:5589]: nice 🌈

Comment (thelostone-mc) [reply app/dashboard/views.py:5589]: Oh wait i think this should be  `polkadot_ext`and `binance_ext`!
Derp derp typo mistake

File: app/dashboard/templates/dashboard/hackathon/new_bounty.html [hunk 6]
```diff
@@ -171,7 +171,7 @@ <h1 class="text-center">Fund Prize</h1>
                     </label>
                   {% endif %}
 
-                  <label class="btn btn-radio chain-btn d-flex align-items-center mr-2 mb-2 font-weight-bold py-2 px-4" :class="{'active': chainId === 58}">
+                  <label class="btn btn-radio chain-btn d-flex align-items-center mr-2 mb-2 font-weight-bold py-2 px-4" :class="{'active': chainId === '58'}">
```
Comment (danlipert) [inline app/dashboard/templates/dashboard/hackathon/new_bounty.html:174]: whats up with this change?

File: app/dashboard/sync/celo.py [hunk 7]
```diff
@@ -63,6 +63,5 @@ def sync_celo_payout(fulfillment):
             fulfillment.payout_status = 'done'
             fulfillment.accepted_on = timezone.now()
             fulfillment.accepted = True
+            fulfillment.save()
             record_payout_activity(fulfillment)
-
-        fulfillment.save()
```
Comment (thelostone-mc) [inline app/dashboard/sync/celo.py:68]: why was this removed ?
This means L58 would not get saved if the txn is in pending

File: app/dashboard/templates/dashboard/hackathon/new_bounty.html (hunk 6 above)
Comment (chibie) [reply app/dashboard/templates/dashboard/hackathon/new_bounty.html:178]: @danlipert when polkadot is clicked, it doesn't appear selected (with blue border). this happens because chainId is a string not integer. '58' fixed that.

Comment (danlipert) [reply app/dashboard/templates/dashboard/hackathon/new_bounty.html:178]: amazing!

Comment (danlipert) [reply app/dashboard/templates/dashboard/hackathon/new_bounty.html:178]: If you make extra fixes like this, I would say either stick it in a different branch OR be sure to note it in the PR description so everyone knows theres some extra goodness in this PR 👍

**Comment by aamustapha** [comment]
An issue we're currently facing is that: the BinanceChain extension doesn't inject the BinanceChain object before our script is loaded. 

We've tried 
- defer script load
- $(document).ready(function () {...})

but didn't work

We're considering having a watcher on the window instance which does our bidding (lol) whenever the window object changes.
Do you see any potential bottleneck with this? @octavioamu @danlipert @thelostone-mc

File: app/dashboard/templates/dashboard/hackathon/new_bounty.html (hunk 6 above)
Comment (chibie) [reply app/dashboard/templates/dashboard/hackathon/new_bounty.html:178]: that's right... a different branch seems cool... since this pr may take a while to be merged

File: app/dashboard/sync/celo.py (hunk 7 above)
Comment (chibie) [reply app/dashboard/sync/celo.py:68]: i noticed it was same way on etc and eth... should i revert this? @thelostone-mc

File: app/dashboard/sync/binance.py (hunk 2 above)
Comment (chibie) [reply app/dashboard/sync/binance.py]: @thelostone-mc

**update:** we reverted to browser header after the smart chain explorer kept bouncing requests with error 403.

File: app/dashboard/sync/celo.py (hunk 7 above)
Comment (thelostone-mc) [reply app/dashboard/sync/celo.py:68]: yup ! I'd say add it to ETC as well

**Comment by thelostone-mc** [comment]
@aamustapha i'm assuming we use the changes only on bounty oayout flow right ? So can we load
- extension script and follow up JS script 
- have the init code with the extension kick in only when the funder clicks on payout ! (harmony integration is done like that)

File: app/assets/v2/images/chains/binance.svg [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+        <path id="Path_93" data-name="Path 93" d="M350.786,91.088l-1.746-1.769a6.066,6.066,0,0,1-4.443,2.012,5.52,5.52,0,0,1,0-11.009,6.154,6.154,0,0,1,4.333,1.923l1.746-2.012a8.066,8.066,0,0,0-6.057-2.41,8,8,0,0,0-.111,16A7.984,7.984,0,0,0,350.786,91.088Z" transform="translate(-181.601 -41.983)" fill="#f0b90b"/>
+        <path id="Path_94" data-name="Path 94" d="M382.706,87.353v6.521h2.719V78.4h-2.719v6.433h-7.383V78.4H372.6V93.874h2.719V87.353Z" transform="translate(-201.002 -42.293)" fill="#f0b90b"/>
+        <path id="Path_95" data-name="Path 95" d="M414.561,78.16h-2.52l-6.809,15.585h2.785l1.592-3.736h7.317l1.57,3.736h2.874Zm1.371,9.439H410.6l2.653-6.19Z" transform="translate(-218.604 -42.164)" fill="#f0b90b"/>
+        <path id="Path_96" data-name="Path 96" d="M445.41,78.4V93.874h2.719V78.4Z" transform="translate(-240.278 -42.293)" fill="#f0b90b"/>
+        <path id="Path_97" data-name="Path 97" d="M470.158,89.1l-8.29-10.7h-2.52V93.874h2.675V82.866l8.533,11.009h2.277V78.4h-2.675Z" transform="translate(-247.797 -42.293)" fill="#f0b90b"/>
+      </g>
+    </g>
+  </g>
+</svg>
```
Comment (thelostone-mc) [inline app/assets/v2/images/chains/binance.svg:24]: can we run svgo on this to compress this ?

File: app/assets/v2/js/pages/hackathon_new_bounty.js [hunk 9]
```diff
@@ -106,6 +106,11 @@ Vue.mixin({
           break;
         }
 
+        case '56': {
+          // INITIALIZE BINANCE WALLET EXTENSION HERE
```
Comment (thelostone-mc) [inline app/assets/v2/js/pages/hackathon_new_bounty.js:110]: i think we can skip the init here unless it's really needed

File: app/dashboard/tests/test_binance_sync.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+            bounty=bounty,
+            payout_tx_id='0xc32f5ad8a1dec9e0ae67f0f55f772ea752e7e032b62b1cdaf8d392e12c66e919'
+        )
+
+    @mock.patch("") 
+    def test_get_binance_txn_status(self, mocked_requests):
+        pass
+
+    def test_sync_binance_payout(self):
+        pass
```
Comment (thelostone-mc) [inline app/dashboard/tests/test_binance_sync.py:79]: ^ is this still a todo section ?

File: app/assets/v2/images/chains/binance.svg (hunk 8 above)
Comment (chibie) [reply app/assets/v2/images/chains/binance.svg]: done!

**Review by danlipert — APPROVED** [review]

**Review by thelostone-mc — APPROVED** [review]
